controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java

```diff
@@ -0,0 +1,417 @@
+        if (value.isDefined()) {
+            ModelNode namedNode = value.get(name);
+            if (namedNode != null && namedNode.isDefined()) {
+                return namedNode.asString();
+            }
+            return null;
+        }
+        return null;
+    }
+
+    /**
+     * Replace injection with new one referencing the same {@link org.wildfly.security.credential.store.CredentialStore} but
+     * based of new values of {@link CredentialReference}
+     * @param injectedCredentialStoreClient {@link InjectedValue} to replace the credential reference
+     * @param credentialReference new credential reference
+     * @throws ClassNotFoundException when credential reference holding credential type which cannot be resolved using current providers
+     */
+    public static void reinjectCredentialStoreClient(InjectedValue<CredentialStoreClient> injectedCredentialStoreClient,
+            CredentialReference credentialReference) throws ClassNotFoundException {
+
+        CredentialStoreClient originalCredentialStoreClient = injectedCredentialStoreClient.getOptionalValue();
+        final CredentialStoreClient updatedCredentialStoreClient;
+        if (originalCredentialStoreClient != null) {
+            updatedCredentialStoreClient =
+                    credentialReference.getCredentialType() != null
+                            ?
+                            new CredentialStoreClient(
+                                    originalCredentialStoreClient.getCredentialStore(),
+                                    credentialReference.getCredentialStoreName(),
+                                    credentialReference.getAlias(),
+                                    credentialReference.getCredentialType())
+                            :
+                            new CredentialStoreClient(
+                                    originalCredentialStoreClient.getCredentialStore(),
+                                    credentialReference.getCredentialStoreName(),
+                                    credentialReference.getAlias());
+        } else {
+            final CredentialStore credentialStore = new ClearTextCredentialStore(credentialReference);
+            updatedCredentialStoreClient = new CredentialStoreClient(credentialStore, CredentialReference.class.getName(), null);
+        }
+        injectedCredentialStoreClient.setValue(() -> updatedCredentialStoreClient);
+    }
+
+    /**
+     * Marshall the value from {@code credentialReferenceModelNode} as an xml element into the given {@code writer}.
+     * @param credentialReferenceModelNode he model, a non-null node of {@link org.jboss.dmr.ModelType#OBJECT}.
+     * @param writer stream writer to use for writing the attribute
+     * @throws XMLStreamException if thrown by {@code writer}
+     */
+    public static void marshallAsElement(ModelNode credentialReferenceModelNode, XMLExtendedStreamWriter writer) throws XMLStreamException {
```

Comment:
I'm being lazy and not researching here, so apologies if this is silly.

ObjectTypeAttributeDefinition.Builder exposes setAttributeMarshaller and setAttributeParser. If the expected uses of this method and readCredentialReference can be covered by using that to set up credentialReferenceAttributeDefinition, and then the callers just use getAttributeDefinition(), then these two methods don't need to be public.

If possible I'd much prefer to not have public utility code for parsing and marshalling proliferating around.